src/resolver.rs

```diff
@@ -1439,57 +1312,211 @@ impl<'i> TypeAnnotator<'i> {
+            base.and_then(|it| {
+                self.annotation_map.get_type(it, self.index).map(|it| it.get_name().to_string())
+            }),
+        ) {
+            (ReferenceAccess::Member(reference), qualifier) => {
+                // uppdate the context's const field
+                let new_ctx = base.map(|base| ctx.with_const(self.is_const_reference(base, ctx)));
+                let new_ctx = new_ctx.as_ref().unwrap_or(ctx);
+
+                if let Some(annotation) =
+                    self.resolve_reference_expression(reference.as_ref(), qualifier.as_deref(), new_ctx)
+                {
+                    self.annotate(stmt, annotation.clone());
+                    self.annotate(reference, annotation);
+                    // if this was a vla, we update a typehint
+                    if self.annotation_map.get_type(stmt, self.index).filter(|it| it.is_vla()).is_some() {
+                        self.annotate_vla_hint(new_ctx, stmt);
+                    }
+                }
+            }
+            (ReferenceAccess::Cast(target), Some(qualifier)) => {
+                // STRING#"abc"
+                //  base
+                if let Some(base_type) = base.and_then(|base| self.annotation_map.get_type(base, self.index))
+                {
+                    // if base is an enum, we need to look for members of this specific enum
+                    let optional_enum_qualifier = Some(qualifier.as_str()).filter(|_| base_type.is_enum());
+                    if ctx.is_in_a_body() {
+                        accept_cast_string_literal(&mut self.string_literals, base_type, target);
+                    }
+
+                    if let Some(annotation) = self.resolve_reference_expression(
+                        target,
+                        optional_enum_qualifier,
+                        &ctx.with_resolving_strategy(vec![ResolvingScope::Variable]),
+                    ) {
+                        self.annotate(target.as_ref(), annotation);
+                        self.annotate(stmt, StatementAnnotation::value(qualifier.as_str()));
+
+                        if let AstStatement::Literal { .. } = target.as_ref() {
+                            // treate casted literals as the casted type
+                            self.annotate(target.as_ref(), StatementAnnotation::value(qualifier.as_str()));
+                        }
+                    }
+                }
+            }
+            (ReferenceAccess::Index(_), Some(base)) => {
+                if let Some(inner_type) = self
+                    .index
+                    .find_effective_type_info(base.as_str())
+                    .and_then(|t| t.get_inner_array_type_name())
+                    .and_then(|it| self.index.find_effective_type_by_name(it).map(|it| it.get_name()))
+                {
+                    self.annotate(stmt, StatementAnnotation::value(inner_type))
+                }
+            }
+            (ReferenceAccess::Deref, _) => {
```

Comment:
ok I remember now why I didnt use this here. Its because the annotation_map.get_type(..) considers self.index AND the internal index in the annotation map that carries new pointer-types that just got created a second ago. There is a test checking for this situation. So I gonna leave this here.